assembly/src/compile.rs

```diff
@@ -0,0 +1,306 @@
+    sync::Arc,
+    vec::Vec,
+};
+
+use crate::{
+    ast::{Module, ModuleKind},
+    diagnostics::{IntoDiagnostic, NamedSource, Report, SourceCode, SourceFile, WrapErr},
+    library::{LibraryNamespace, LibraryPath},
+};
+
+// COMPILE OPTIONS
+// ================================================================================================
+
+/// The set of options which can be used to control the behavior of the [Compile] trait.
+#[derive(Debug, Clone)]
+pub struct Options {
+    /// The kind of [Module] to compile.
+    ///
+    /// The default kind is executable.
+    pub kind: ModuleKind,
+    /// When true, promote warning diagnostics to errors
+    pub warnings_as_errors: bool,
+    /// The name to give the compiled [Module]
+    ///
+    /// This option overrides `namespace`.
+    ///
+    /// If unset, and there is no name associated with the item being compiled (e.g. a file path)
+    /// then the path will consist of just a namespace; using the value of `namespace` if provided,
+    /// or deriving one from `kind`.
+    pub path: Option<LibraryPath>,
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Self {
+            kind: ModuleKind::Executable,
+            warnings_as_errors: false,
+            path: None,
+        }
+    }
+}
+impl Options {
+    /// Configure a set of [Options] to compile a [Module] with the given `kind` and `path`.
+    ///
+    /// This is primarily useful when compiling a module from source code that has no meaningful
+    /// [LibraryPath] associated with it, such as when compiling from a `str`. This will override
+    /// the default name derived from the given [ModuleKind].
+    pub fn new<P, E>(kind: ModuleKind, path: P) -> Result<Self, E>
+    where
+        P: TryInto<LibraryPath, Error = E>,
+    {
+        let path = path.try_into()?;
+        Ok(Self {
+            kind,
+            path: Some(path),
+            ..Default::default()
+        })
+    }
+
+    /// Get the default [Options] for compiling a library module.
+    pub fn for_library() -> Self {
+        Self {
+            kind: ModuleKind::Library,
+            ..Default::default()
+        }
+    }
+
+    /// Get the default [Options] for compiling a kernel module.
+    pub fn for_kernel() -> Self {
+        Self {
+            kind: ModuleKind::Kernel,
+            ..Default::default()
+        }
+    }
+}
+
+// COMPILE TRAIT
+// ================================================================================================
+
+/// This trait is meant to be implemented by any type that can be compiled to a [Module],
+/// to allow methods which expect a [Module] to accept things like:
+///
+/// * A [Module] which was previously compiled or read from a [crate::Library].
+/// * A string representing the source code of a [Module].
+/// * A path to a file containing the source code of a [Module].
+/// * A vector of [Form]s comprising the contents of a [Module].
+pub trait Compile: Sized {
```

Comment:
I guess to take a step back, the purpose of this trait is primarily the following:

* Avoid duplication in the API surface of `Assembler`. This trait replaces the many variations of `add_module` and the like that we had before to support providing modules in different source forms or in AST form. The vast majority of users of these APIs were passing modules as strings, file paths, or AST, and aside from the last of those (which acted as a kind of a lower-level API for the others), these APIs were parsing/validating the `Module` and then passing it to the AST-based API.
* Ergonomics. We want to be able to do things like call `assembler.add_module` with anything that derefs to `str`, `Path`, or a `Module` and have it "just work". Existing assembler-related code makes liberal use of this; especially the test suite which used/uses this capability quite extensively to express tests succinctly.
* Consistency. We can ensure that "compiling" a `Module` is always done the same way, and maintained in a single place, rather than spread across many different areas of the code. Previously, this was re-implemented, by hand, in every single place where you wanted to get a `Module` from a string/file/whatever.

The trait essentially encapsulates parsing a `Module` and performing semantic analysis on it. This process was being referred to as compilation previously, so I kept that terminology here (though personally I think that is maybe a bit confusing, since the only thing that actually resembles compilation here is when we lower to MAST during assembly).

With that in mind, if we were to rename this trait, I think it's important for the name adhere to the following:

* The trait name should be reflected in its API (or vice versa, however you prefer to think about it) e.g. `Compile` provides a `compile` method. This is generally the case in Rust as well with traits that provide a single method/set of related methods, e.g. `Write`, `Hash`, `From`, `AsRef`, and so on.
* A trait that represents an action, like conversion, should use the imperative: e.g. `ToString`, `FromStr`, `TryFrom`.
* Ideally, the semantics of the trait are reflected in the name. In other words, it should be obvious if the trait is intended to be an intrinsic property of the implementor, or a fallible/optional capability of the implementor, e.g. the distinction between `From` and `TryFrom`.

So putting all of that together, `Compile` is what I landed on, since it hits all of the main points I listed. I also felt that, since it was primarily going to be appearing in the public API of the `Assembler`, passing a value that implements `Compile` felt like it meshed well with how that value was going to be used in by the assembler, even if the trait itself is more about parsing/analysis than the full end-to-end compilation pipeline.

That said, if I were to pick an alternative way of representing this, here are some ideas

* Rename the trait to `Parse`, add an associated type `Value`, and then use `impl Parse<Value = Module>` instead of `impl Compile`. This has the benefit of being extensible to other syntax elements other than `Module` that we might want to be able to parse independently. I like that it reflects exactly what functionality has to be implemented.
* Get rid of the `Compile` trait entirely, and instead create a new type that represents the combined source code and compilation/parser options as a single value. We can then implement `From`/`TryFrom`/etc. for types which are trivially/fallibly convertible to this new type, as well as provide a builder API for that type so that it can be constructed more iteratively. The existing uses of `impl Compile` would then become either `impl TryInto<T>` or just `T`, where `T` is this new type. Not 100% sure what to call this proposed type, but `ModuleSource` would be more appropriate here than as a trait name IMO.
* Remove the ergonomic APIs from the `Assembler` entirely, and require always adding modules in their AST form. This keeps the `Assembler` API simple, but does push responsibility for obtaining a `Module` on to downstream users of the assembler. This would be a more invasive breaking change, but would simplify things considerably.

What's your take?

Reply:
I like the new approach (i.e., using `impl Compile`), my comment was mostly about the trait name. Your arguments re naming make sense - but I still think that something like `ModuleSource` might be a more intuitive name. Having said this, this is not a strong preference by any means and I'm fine with keeping `Compile` name if you feel strongly about it.